# CPP

[C++ 支持 - ESP32 - — ESP-IDF 编程指南 v5.2.3 文档](https://docs.espressif.com/projects/esp-idf/zh_CN/v5.2.3/esp32/api-guides/cplusplus.html)

ESP-IDF 支持以下 C++ 功能：

- [异常处理](https://docs.espressif.com/projects/esp-idf/zh_CN/v5.2.3/esp32/api-guides/cplusplus.html#cplusplus-exceptions)
- [多线程](https://docs.espressif.com/projects/esp-idf/zh_CN/v5.2.3/esp32/api-guides/cplusplus.html#cplusplus-multithreading)
- [运行时类型信息 (RTTI)](https://docs.espressif.com/projects/esp-idf/zh_CN/v5.2.3/esp32/api-guides/cplusplus.html#cplusplus-rtti)
- [线程局部存储](https://docs.espressif.com/projects/esp-idf/zh_CN/v5.2.3/esp32/api-guides/thread-local-storage.html) （`thread_local` 关键字）
- 除部分 [限制](https://docs.espressif.com/projects/esp-idf/zh_CN/v5.2.3/esp32/api-guides/cplusplus.html#cplusplus-limitations)，所有由 GCC 实现的 C++ 功能均受支持。有关由 GCC 所实现功能的详细信息，请参阅 [GCC 文档](https://gcc.gnu.org/projects/cxx-status.html)。

## 异常处理

### 异常捕获

[C++ 异常处理 | 菜鸟教程](https://www.runoob.com/cplusplus/cpp-exceptions-handling.html)

异常是程序在执行期间产生的问题。C++ 异常是指在程序运行时发生的特殊情况，比如尝试除以零的操作。

异常提供了一种转移程序控制权的方式。C++ 异常处理涉及到三个关键字：**try、catch、throw**。

- **throw:** 当问题出现时，程序会抛出一个异常。这是通过使用 **throw** 关键字来完成的。
- **catch:** 在您想要处理问题的地方，通过异常处理程序捕获异常。**catch** 关键字用于捕获异常。
- **try:** **try** 块中的代码标识将被激活的特定异常。它后面通常跟着一个或多个 catch 块。

```cpp
#include <iostream>
using namespace std;

double division(int a, int b) {
	if (b == 0) {
		throw "Division by zero condition!";
	}
	return (a / b);
}

int main() {
	int x = 50;
	int y = 0;
	double z = 0;

	try {
		z = division(x, y);
		cout << z << endl;
	} catch (const char* msg) {
		cerr << msg << endl;
	}	

	return 0;
}
```

如果`try`块可能抛出不同类型的异常，可以使用多个`catch`块来捕获不同类型的异常。

```cpp
try {
	// 可能抛出不同类型的异常
} catch (DataType1& e1) {
	// 处理DataType1类型的异常
} catch (DataType2& e2) {
	// 处理DataType2类型的异常
} catch (DataType3& e3) {
	// 处理DataType3类型的异常
} catch (...) {
	// 处理所有其他类型的异常
}
```

cpp专门使用一组`<exception>`类用于返回错误

![image-20250122233108235](https://picture-01-1316374204.cos.ap-beijing.myqcloud.com/picture/202501222331375.png)

| 异常                   | 描述                                                         |
| :--------------------- | :----------------------------------------------------------- |
| **std::exception**     | 该异常是所有标准 C++ 异常的父类。                            |
| std::bad_alloc         | 该异常可以通过 **new** 抛出。                                |
| std::bad_cast          | 该异常可以通过 **dynamic_cast** 抛出。                       |
| std::bad_typeid        | 该异常可以通过 **typeid** 抛出。                             |
| std::bad_exception     | 这在处理 C++ 程序中无法预期的异常时非常有用。                |
| **std::logic_error**   | 理论上可以通过读取代码来检测到的异常。                       |
| std::domain_error      | 当使用了一个无效的数学域时，会抛出该异常。                   |
| std::invalid_argument  | 当使用了无效的参数时，会抛出该异常。                         |
| std::length_error      | 当创建了太长的 std::string 时，会抛出该异常。                |
| std::out_of_range      | 该异常可以通过方法抛出，例如 std::vector 和 std::bitset<>::operator[]()。 |
| **std::runtime_error** | 理论上不可以通过读取代码来检测到的异常。                     |
| std::overflow_error    | 当发生数学上溢时，会抛出该异常。                             |
| std::range_error       | 当尝试存储超出范围的值时，会抛出该异常。                     |
| std::underflow_error   | 当发生数学下溢时，会抛出该异常。                             |

实际使用的时候可以使用`.what()`获取实际的错误, 自定义的时候也需要重写这一个函数

```cpp
class bad_hmean : public std::exception
{
public:
	const char* what()
	{
		return "bad arguments to hmean()";
	}
	// ...
};

try{
	//...
}
catch(std::exception& e)
{
	cout << e.what() << endl;
	// ...
}
```

也可以改写描述

```cpp
//可以使用()来传递参数, 参数类型必须是const char*或者std::string, 这一个参数会被传递给std::runtime_error的构造函数, 作为异常的描述信息, 使用what()函数可以获取这个描述信息
throw std::runtime_error("Runtime error occurred"); 
```



### 异常类型

`#include <optional>`

使用这一个头文件的时候可以使用类型`std::optional<int>`这一个类型的范围是[INT_MIN, INT_MAX] | {nullopt}, 所以可以返回`std::nullopt`表示失败

判断的时候可以使用`.has_value()`进行判断, 使用`.value()`获取数值, 也可以直接判断返回值`if(ret)`是不是为空

> 使用`*ret`的时候不会进行检查, 相当于使用解引用一个NULL

这一种方式只有一种输出, 如果使用返回值为负数的情况会使得输出的范围减小, 所以也可以使用一个指针参数作为返回

+ 使用一个全局变量进行记录

使用头文件`#include "errno.h"`, 使用全局变量errno, 这是每一个线程独属的变量

在这个头文件里面定义了一系列的错误

使用`#include <cstring>`里面的`const char *str = strerror(ENOENT);`可以获取描述, 使用`setlocale(LC_ALL, "zh_CN.UTF-8");`返回的是中文

也可以使用`perror("你的提示")`进行输出

> 使用这一个的时候自定义的程度比较差, 所以可以自己建立一个枚举类型

## 多任务

Mutex 又称互斥量，C++ 11中与 Mutex 相关的类（包括锁类型）和函数都声明在 `<mutex>` 头文件中，所以如果你需要使用 std::mutex，就必须包含` <mutex> `头文件。

Mutex 系列类(四种)

- std::mutex，最基本的 Mutex 类, 不支持递归。
- std::recursive_mutex，递归 Mutex 类。
- std::time_mutex，定时 Mutex 类。
- std::recursive_timed_mutex，定时递归 Mutex 类。

Lock 类（两种）

- std::lock_guard，与 Mutex RAII 相关，方便线程对互斥量上锁。
- std::unique_lock，与 Mutex RAII 相关，方便线程对互斥量上锁，但提供了更好的上锁和解锁控制。

其他类型

- std::once_flag
- std::adopt_lock_t
- std::defer_lock_t
- std::try_to_lock_t

函数

- std::try_lock，尝试同时对多个互斥量上锁。
- std::lock，可以同时对多个互斥量上锁。
- std::call_once，如果多个线程需要同时调用某个函数，call_once 可以保证多个线程对该函数只调用一次。

### std::mutex

- 构造函数，std::mutex不允许拷贝构造，也不允许 move 拷贝，最初产生的 mutex 对象是处于 unlocked 状态的。
- lock()，调用线程将锁住该互斥量。线程调用该函数会发生下面 3 种情况：(1). 如果该互斥量当前没有被锁住，则调用线程将该互斥量锁住，直到调用 unlock之前，该线程一直拥有该锁。(2). 如果当前互斥量被其他线程锁住，则当前的调用线程被阻塞住。(3). 如果当前互斥量被当前调用线程锁住，则会产生死锁(deadlock)。
- unlock()， 解锁，释放对互斥量的所有权。
- try_lock()，尝试锁住互斥量，如果互斥量被其他线程占有，则当前线程也不会被阻塞。线程调用该函数也会出现下面 3 种情况，(1). 如果当前互斥量没有被其他线程占有，则该线程锁住互斥量，直到该线程调用 unlock 释放互斥量。(2). 如果当前互斥量被其他线程锁住，则当前调用线程返回 false，而并不会被阻塞掉。(3). 如果当前互斥量被当前调用线程锁住，则会产生死锁(deadlock)。

### std::recursive_mutex

std::recursive_mutex 允许同一个线程对互斥量多次上锁（即递归上锁）

### std::time_mutex

std::time_mutex 比 std::mutex 多了两个成员函数，try_lock_for()，try_lock_until()。

try_lock_for 函数接受一个时间范围，表示在这一段时间范围之内线程如果没有获得锁则被阻塞住（与 std::mutex 的 try_lock() 不同，try_lock 如果被调用时没有获得锁则直接返回 false），如果在此期间其他线程释放了锁，则该线程可以获得对互斥量的锁，如果超时（即在指定时间内还是没有获得锁），则返回 false。

###  std::lock_guard

一种自动管理 std::mutex 锁的封装器，使用 RAII 风格，确保在作用域结束时自动释放锁。

```cpp
#include <mutex>

std::mutex mtx;

void function() {
    std::lock_guard<std::mutex> lock(mtx);
    // 访问共享资源
}
```

### std::unique_lock

提供比 std::lock_guard 更灵活的锁管理，可以手动释放和重新获得锁，还支持定时锁定。

```cpp
#include <mutex>
#include <chrono>

std::mutex mtx;

void function() {
    std::unique_lock<std::mutex> lock(mtx);
    // 访问共享资源
   
    // 可以手动释放锁
    lock.unlock();
   
    // 可以重新获得锁
    lock.lock();
   
    // 可以进行定时锁定
    if (lock.try_lock_for(std::chrono::seconds(1))) {
        // 成功获得锁
    }
}
```

### 原子变量

[C++ 中的原子变量（std::atomic）使用指南_std::atomic-CSDN博客](https://blog.csdn.net/qq_46017342/article/details/132838649)

原子变量（`std::atomic`）是C++中用于多线程编程的强大工具之一。它们提供了一种线程安全的方式来访问和修改共享数据，而无需使用显式的互斥锁。

`std::atomic`支持各种数据类型，如整数、布尔值、指针等。您可以创建`std::atomic`对象，并使用原子操作来读取和修改它们的值。

```cpp
std::atomic<int> atomicInt(0);
std::atomic<bool> atomicBool(true);
```

+ 读取值

要读取std::atomic变量的值，可以使用load函数：

```cpp
int value = atomicInt.load();
bool flag = atomicBool.load();
```

+ 修改值

要修改std::atomic变量的值，可以使用store函数：

```cpp
atomicInt.store(42);
atomicBool.store(false);
```

### std::condition_variable

用于在多线程程序中进行线程间的同步。它允许一个或多个线程在某个条件为真之前挂起（等待），直到另一个线程修改了条件并通知 std::condition_variable

`std::condition_variable` 需要与 `std::unique_lock<std::mutex>` 一起使用，以便在等待条件时释放互斥锁，并在条件满足后重新获取互斥锁。它提供了 `wait`、`notify_one` 和 `notify_all` 方法来实现线程间的等待和通知机制。

- `wait` 方法使当前线程阻塞，直到条件变量被另一个线程通过 *notify_one* 或 `notify_all` 方法唤醒。
- `notify_one` 方法唤醒一个等待条件变量的线程。
- `notify_all` 方法唤醒所有等待条件变量的线程。